src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.ValueTaskSource.cs

```diff
@@ -29,37 +29,36 @@ private unsafe class ValueTaskSource : IValueTaskSource<int>, IValueTaskSource
-                // which has a PreAllocatedOverlapped with the memory already pinned.  Otherwise, we use the derived
-                // MemoryAwaitableProvider, which Retains the memory, which will result in less pinning in the case
-                // where the underlying memory is backed by pre-pinned buffers.
-                return preallocatedOverlapped != null &&
-                       MemoryMarshal.TryGetArray(memory, out ArraySegment<byte> buffer) &&
-                       preallocatedOverlapped.IsUserObject(buffer.Array) ?
-                            new ValueTaskSource(strategy, preallocatedOverlapped, buffer.Array) :
-                            new MemoryValueTaskSource(strategy, memory);
-            }
 
-            protected ValueTaskSource(
-                AsyncWindowsFileStreamStrategy strategy,
-                PreAllocatedOverlapped? preallocatedOverlapped,
-                byte[]? bytes)
+            internal ValueTaskSource(AsyncWindowsFileStreamStrategy strategy, byte[]? internalBuffer)
             {
                 _strategy = strategy;
                 _result = TaskSourceCodes.NoResult;
-
                 _source = default;
                 _source.RunContinuationsAsynchronously = true;
 
-                _overlapped = bytes != null &&
-                              _strategy.CompareExchangeCurrentOverlappedOwner(this, null) == null ?
-                              _strategy._fileHandle.ThreadPoolBinding!.AllocateNativeOverlapped(preallocatedOverlapped!) : // allocated when buffer was created, and buffer is non-null
-                              _strategy._fileHandle.ThreadPoolBinding!.AllocateNativeOverlapped(s_ioCallback, this, bytes);
+                if (internalBuffer != null)
+                {
+                    _preallocatedOverlapped = new PreAllocatedOverlapped(s_ioCallback, this, internalBuffer);
+                }
+            }
+
+            internal void Configure(ReadOnlyMemory<byte> memory)
+            {
+                // The array of preallocatedOverlapped is the same as the memory array when flushing in buffered mode,
+                // because the ReadOnlyMemory object passed to WriteAsync/ReadAsync is the internal buffer that we need
+                // to write to disk at the end
+                if (_preallocatedOverlapped != null &&
+                   MemoryMarshal.TryGetArray(memory, out ArraySegment<byte> arrSegment) &&
+                   _preallocatedOverlapped.IsUserObject(arrSegment.Array))
+                {
+                    _overlapped = _strategy._fileHandle.ThreadPoolBinding!.AllocateNativeOverlapped(_preallocatedOverlapped);
+                }
+                else
+                {
+                    _handle = memory.Pin();
+                    _overlapped = _strategy._fileHandle.ThreadPoolBinding!.AllocateNativeOverlapped(s_ioCallback, this, null);
+                }
```

Comment:
should we also reset the `_result` field as it might have a value from the previous usage?

```suggestion
                }

                _result = TaskSourceCodes.NoResult;
```